core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -719,7 +719,20 @@ object SparkSubmit extends CommandLineUtils {
       printWarning("Subclasses of scala.App may not work correctly. Use a main() method instead.")
     }
 
-    val mainMethod = mainClass.getMethod("main", new Array[String](0).getClass)
+    val isSparkApp = mainClass.getMethods().filter{_.getName() == "sparkMain"}.length > 0
+
+    val childSparkConf = sysProps.filter{ p => p._1.startsWith("spark.")}.toMap
+    if (!isSparkApp || !threadEnabled) {
+      sysProps.foreach { case (key, value) =>
+        System.setProperty(key, value)
+      }
+    }
+
+    val mainMethod = if (isSparkApp) {
+      mainClass.getMethods().filter{_.getName() == "sparkMain"}(0)
```

Comment:
we have already decided above if its a spark app so we don't really need to getMethods and filter again.